nativelib/src/main/resources/gc/immix/Heap.c

```diff
@@ -290,33 +315,63 @@ bool Heap_shouldGrow(Heap *heap) {
            4 * unavailableBlockCount > blockCount;
 }
 
-void Heap_Recycle(Heap *heap) {
-    Allocator_Clear(&allocator);
-    LargeAllocator_Clear(&largeAllocator);
-    BlockAllocator_Clear(&blockAllocator);
+void Heap_sweep(Heap *heap, uint32_t maxCount) {
+    uint64_t start_ns, end_ns;
+    Stats *stats = heap->stats;
+    if (stats != NULL) {
+        start_ns = scalanative_nano_time();
+    }
+
+    uint32_t startIdx = heap->sweep.cursor;
+    uint32_t limitIdx = startIdx + maxCount;
+    heap->sweep.cursor = limitIdx;
+    uint32_t blockCount = heap->blockCount;
+    if (limitIdx > blockCount) {
+        limitIdx = blockCount;
+    }
 
-    BlockMeta *current = (BlockMeta *)heap->blockMetaStart;
-    word_t *currentBlockStart = heap->heapStart;
-    LineMeta *lineMetas = (LineMeta *)heap->lineMetaStart;
-    word_t *end = heap->blockMetaEnd;
-    while ((word_t *)current < end) {
+    BlockMeta *current = BlockMeta_GetFromIndex(heap->blockMetaStart, startIdx);
+    BlockMeta *limit = BlockMeta_GetFromIndex(heap->blockMetaStart, limitIdx);
+    word_t *currentBlockStart = Block_GetStartFromIndex(heap->heapStart, startIdx);
+    LineMeta *lineMetas = Line_getFromBlockIndex(heap->lineMetaStart, startIdx);
+    while (current < limit) {
         int size = 1;
-        assert(!BlockMeta_IsSuperblockMiddle(current));
         if (BlockMeta_IsSimpleBlock(current)) {
             Block_Recycle(&allocator, current, currentBlockStart, lineMetas);
         } else if (BlockMeta_IsSuperblockStart(current)) {
             size = BlockMeta_SuperblockSize(current);
             LargeAllocator_Sweep(&largeAllocator, current, currentBlockStart);
-        } else {
-            assert(BlockMeta_IsFree(current));
+        } else if (BlockMeta_IsFree(current)) {
             BlockAllocator_AddFreeBlocks(&blockAllocator, current, 1);
         }
+        // ignore superblock middle blocks, that superblock will be swept by someone else
         assert(size > 0);
         current += size;
         currentBlockStart += WORDS_IN_BLOCK * size;
         lineMetas += LINE_COUNT * size;
     }
+    heap->sweep.cursorDone = limitIdx;
+
+    if (stats != NULL) {
+        end_ns = scalanative_nano_time();
+        Stats_RecordLazySweep(stats, start_ns, end_ns);
+    }
+
+    if (Heap_IsSweepDone(heap)) {
+        Heap_sweepDone(heap);
+    }
+}
+
+void Heap_Recycle(Heap *heap) {
+    Allocator_Clear(&allocator);
+    LargeAllocator_Clear(&largeAllocator);
+    BlockAllocator_Clear(&blockAllocator);
 
+    heap->sweep.cursor = 0;
+    Heap_sweep(heap, heap->blockCount);
```

Comment:
This is actually NOT lazy sweeping.  The line here sweeps the entire heap. The performance improvement probably comes from changes to `BlockList` and `BlockAllocator` that were needed to enable lazy sweeping.

Reply:
I will remove this line and run the benchmarks again.